Lavellt create account

TuckerL-OSU opened the pull request:
checking diff

taylor-jones committed:
Added initial files for report page and started writing SQL report queries

taylor-jones committed:
Udpated AwardTypeController to use userId cookie value instead of param

taylor-jones committed:
Added a few report-like queries to the server and started calling them from report page

tmrietz committed:
Add chart for user award counts
  - Create a simpler, reusable vertical-bar-chart wrapper
  - Integrate with user award counts query and service
  - Add to reports page

tmrietz committed:
Add form input validations
  - Admin control forms and account page form validations
  - Include custom validation for signature/canvas element
  - General cleanup (make admin page header consistent with others)

TuckerL-OSU committed:
init create user form

TuckerL-OSU committed:
pulled

TuckerL-OSU committed:
Added Create User as an unlogged in user
-currently saves the user info.
-properly logs account creating from subjectId 0

TODO:
-Save recovery questions
-frontend form validation on create account
-frontend form validation on forgot password
-fix log support for forgot password change

tmrietz commented on erp-client/src/app/models/new-user.model.ts:
```diff
@@ -0,0 +1,10 @@
+  username: string;
+  password: string;
+  rePassword: string;
+  recoveryQuestions: RecoveryQuestion[];
```
Will I need to go back and add this for her creation by admins?

TuckerL-OSU replied:
I'm not sure what you mean

TuckerL-OSU replied:
Yea..I forgot we need to set recovery questions. Strange for the admin to put those and the persons signature in, but I haven't really thought of any other way to do the admin creating users.

TuckerL-OSU committed:
Added Create User as an unlogged in user
-currently saves the user info.
-Saves recovery questions
-properly logs account creating from subjectId 0

TODO:
-frontend form validation on create account
-frontend form validation on forgot password
-fix log support for forgot password change

TuckerL-OSU committed:
Added Create User as an unlogged in user
-currently saves the user info.
-Saves recovery questions
-properly logs account creating from subjectId 0

TODO:
-frontend form validation on create account
-frontend form validation on forgot password
-fix log support for forgot password change
-add log support for create account

TuckerL-OSU committed:
Merged dev into account creation

tmrietz commented on erp-server/src/main/java/com/ttt/erp/service/RecoveryService.java:
```diff
@@ -48,12 +48,13 @@ public Boolean validateRecoveryQuestions(String username, final List<RecoveryQue
+    public Boolean updateUsersPassword(String username, String password) {
         try {
             UserAccount existing = this.userAccountRepository.findByUsername(username);
             UserAccount modified = existing;
```
```modified``` and ```existing``` are referencing the same object. That's why when you say ```modified.setPassword(password)```, the password on ```existing``` gets changed too and nothing gets logged. If objects are the same then there's of course no diff between them. To demonstrate, I pulled down your code and ran through the password reset.

My code change to print results looked like:
```Java
            System.out.println(modified.toString());
            System.out.println(existing.toString());
            modified.setPassword(password);
            System.out.println(modified.toString());
            System.out.println(existing.toString());
```

And in the console, it looked like:
```
UserAccount{id=2, email='demo@demo.com', username='foo', password='demo2', signature='d123c23f-1801-4002-91c0-d4cb43c4bdf0.png', isAdmin=false', isEnabled=true}
UserAccount{id=2, email='demo@demo.com', username='foo', password='demo2', signature='d123c23f-1801-4002-91c0-d4cb43c4bdf0.png', isAdmin=false', isEnabled=true}
UserAccount{id=2, email='demo@demo.com', username='foo', password='demo', signature='d123c23f-1801-4002-91c0-d4cb43c4bdf0.png', isAdmin=false', isEnabled=true}
UserAccount{id=2, email='demo@demo.com', username='foo', password='demo', signature='d123c23f-1801-4002-91c0-d4cb43c4bdf0.png', isAdmin=false', isEnabled=true}
```
So confirmed that's the root cause.

TuckerL-OSU committed:
Added Create User as an unlogged in user
-currently saves the user info.
-Saves recovery questions
-properly logs account creating from subjectId 0
-added log support for create account

TODO:
-frontend form validation on create account
-frontend form validation on forgot password
-fix log support for forgot password change
--When I step through, in the entity diff, it checks the password is changed, but it also checks for a diff in the awards, questions, and logs fields. But since userRepo.findById and findByUsername, don't return those, so it always sees them as a diff. Then it tries to insert that change of losing everything associted with that id into the log table, which then breaks.
--also happens on update user admin page

TuckerL-OSU committed:
Merge branch 'dev/master' into lavellt-create-account
